src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -2576,18 +2598,162 @@ CalculateNewConnectionCount(WorkerPool *workerPool)
 		/*
 		 * Open enough connections to handle all tasks that are ready, but no more
 		 * than the target pool size.
+		 *
+		 * But before that, we do one more check. If the cost of executing the
+		 * remaining tasks over the existing connections is cheaper than
+		 * establishing new connections and executing the tasks over the new
+		 * connections, we prefer the former.
+		 *
+		 * Note that we are not trying to find the optimal number of connections
+		 * for the remaining tasks here. Our goal is to prevent connection
+		 * establishments that are absolutely unnecessary. In the future, we may
+		 * improve the calculations below to find the optimal number of new
+		 * connections required.
 		 */
 		newConnectionCount = Min(newConnectionsForReadyTasks, maxNewConnectionCount);
-		if (newConnectionCount > 0)
+		if (newConnectionCount > 0 &&
+			UsingExistingConnectionsCheaperThanEstablishingNewConnections(
+				newConnectionCount, workerPool))
```

Comment:
I wonder whether we should give users control over whether or not to use this. Predictability is quite valuable and users sometimes end up enabling force_max_query_parallelization.

Reply:
I added an enum, which I though we could add other alternatives in the future. I can convert it back to a boolean